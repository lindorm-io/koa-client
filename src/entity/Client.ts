import { ClientEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  extra: TObject<any>;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  extra?: TObject<any>;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _extra: TObject<any>;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._extra = options?.extra || {};
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get extra(): TObject<any> {
    return this._extra;
  }
  public set extra(object: TObject<any>) {
    this._extra = object;
    this.addEvent(ClientEvent.EXTRA_CHANGED, { extra: this._extra });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new EntityCreationError("Client");
    }
    this.addEvent(ClientEvent.CREATED, {
      approved: this._approved,
      description: this._description,
      extra: this._extra,
      name: this._name,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
