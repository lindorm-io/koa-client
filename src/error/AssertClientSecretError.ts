import { APIError, HttpStatus } from "@lindorm-io/core";

export class AssertClientSecretError extends APIError {
  constructor(clientId: string, error?: Error) {
    super("Invalid Client Secret", {
      debug: { id: clientId, error },
      statusCode: HttpStatus.ClientError.FORBIDDEN,
    });
  }
}
