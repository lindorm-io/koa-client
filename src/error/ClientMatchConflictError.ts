import { APIError, HttpStatus } from "@lindorm-io/core";

export class ClientMatchConflictError extends APIError {
  constructor(expect: string, actual: string) {
    super("Client Match Conflict", {
      debug: { expect, actual },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
