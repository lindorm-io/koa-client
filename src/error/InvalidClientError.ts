import { APIError, HttpStatus } from "@lindorm-io/core";

export class InvalidClientError extends APIError {
  constructor(clientId: string, error?: Error) {
    super("Invalid Client ID", {
      debug: { id: clientId, error },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
