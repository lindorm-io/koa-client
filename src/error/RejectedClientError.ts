import { APIError, HttpStatus } from "@lindorm-io/core";

export class RejectedClientError extends APIError {
  constructor(clientId: string, error?: Error) {
    super("Client is not approved", {
      debug: { id: clientId, error },
      statusCode: HttpStatus.ClientError.BAD_REQUEST,
    });
  }
}
