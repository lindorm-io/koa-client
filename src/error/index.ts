export * from "./AssertClientSecretError";
export * from "./ClientMatchConflictError";
export * from "./InvalidClientError";
export * from "./RejectedClientError";
