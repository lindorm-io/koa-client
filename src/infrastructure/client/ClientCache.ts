import { CacheBase, ICache, ICacheOptions } from "@lindorm-io/redis";
import { IClient, Client } from "../../entity";
import { schema } from "./schema";

export interface IClientCache extends ICache<Client> {
  create(entity: Client): Promise<Client>;
  update(entity: Client): Promise<Client>;
  find(id: string): Promise<Client>;
  findAll(): Promise<Array<Client>>;
  remove(entity: Client): Promise<void>;
}

export class ClientCache extends CacheBase<Client> implements IClientCache {
  constructor(options: ICacheOptions) {
    super({
      client: options.client,
      entityName: Client.constructor.name,
      expiresInSeconds: options.expiresInSeconds,
      logger: options.logger,
      schema,
    });
  }

  protected createEntity(data: IClient): Client {
    return new Client(data);
  }

  protected getEntityJSON(entity: Client): IClient {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      approved: entity.approved,
      description: entity.description,
      extra: entity.extra,
      name: entity.name,
      secret: entity.secret,
    };
  }

  protected getEntityKey(entity: Client): string {
    return entity.id;
  }

  public async create(entity: Client): Promise<Client> {
    return super.create(entity);
  }

  public async update(entity: Client): Promise<Client> {
    return super.update(entity);
  }

  public async find(id: string): Promise<Client> {
    return super.find(id);
  }

  public async findAll(): Promise<Array<Client>> {
    return super.findAll();
  }

  public async remove(entity: Client): Promise<void> {
    await super.remove(entity);
  }
}
