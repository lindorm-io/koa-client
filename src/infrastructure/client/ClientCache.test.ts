import MockDate from "mockdate";
import { CacheEntityNotFoundError } from "@lindorm-io/redis/dist/error";
import { Client } from "../../entity";
import { ClientCache } from "./ClientCache";
import { IRedisConnectionOptions, RedisConnection, RedisConnectionType } from "@lindorm-io/redis";
import { Logger, LogLevel } from "@lindorm-io/winston";
import { baseHash } from "@lindorm-io/core";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

const redisOptions: IRedisConnectionOptions = {
  type: RedisConnectionType.MEMORY,
  port: 12345,
};

const logger = new Logger({ packageName: "n", packageVersion: "v" });
logger.addConsole(LogLevel.ERROR);

describe("ClientCache", () => {
  let cache: ClientCache;
  let client: Client;

  beforeEach(async () => {
    const redis = new RedisConnection(redisOptions);
    await redis.connect();

    cache = new ClientCache({
      client: redis.getClient(),
      logger,
    });
    client = new Client({
      approved: true,
      description: "description",
      extra: { emailAuthorizationUri: "https://lindorm.io" },
      name: "name",
      secret: baseHash("secret"),
    });
  });

  test("should create", async () => {
    await expect(cache.create(client)).resolves.toMatchSnapshot();
  });

  test("should update", async () => {
    await cache.create(client);

    await expect(cache.update(client)).resolves.toMatchSnapshot();
  });

  test("should find", async () => {
    await cache.create(client);

    await expect(cache.find(client.id)).resolves.toMatchSnapshot();
  });

  test("should remove", async () => {
    await cache.create(client);

    await expect(cache.remove(client)).resolves.toBe(undefined);
    await expect(cache.find(client.id)).rejects.toStrictEqual(expect.any(CacheEntityNotFoundError));
  });
});
