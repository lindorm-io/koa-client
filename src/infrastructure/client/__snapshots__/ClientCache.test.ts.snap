// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientCache should create 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`ClientCache should find 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`ClientCache should update 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;
